From a626ae0b09750f0ff6bdfaab348579a378723e36 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 19 Nov 2012 21:05:47 -0600
Subject: [PATCH] Fix players not getting correctly attached to entities -
 fixes BUKKIT-2993

Often times when the EntityTracker is spawning entities for a player because they have entered that player's visible area, it will sometimes spawn the player first and sometimes spawn the vehicle first.

In order to fix both issues, we need to make sure we're accounting for both cases. Because the client doesn't validate Packet39AttachEntity or care if the entity given doesn't exist, sending an extra packet is harmless.
---
 .../net/minecraft/server/EntityTrackerEntry.java   |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index acb3637..19260e3 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -315,9 +315,13 @@ public class EntityTrackerEntry {
                         entityplayer.netServerHandler.sendPacket(new Packet28EntityVelocity(this.tracker.id, this.tracker.motX, this.tracker.motY, this.tracker.motZ));
                     }
 
-                    if (this.tracker.vehicle != null) {
+                    // CraftBukkit start
+                    if (this.tracker.vehicle != null && this.tracker.id > this.tracker.vehicle.id) {
                         entityplayer.netServerHandler.sendPacket(new Packet39AttachEntity(this.tracker, this.tracker.vehicle));
+                    } else if (this.tracker.passenger != null && this.tracker.id > this.tracker.passenger.id) {
+                        entityplayer.netServerHandler.sendPacket(new Packet39AttachEntity(this.tracker.passenger, this.tracker));
                     }
+                    // CraftBukkit end
 
                     if (this.tracker instanceof EntityLiving) {
                         for (int i = 0; i < 5; ++i) {
-- 
1.7.4.4

